src/lib/io/sol-iio.c

```diff
@@ -42,11 +42,15 @@
 #include <sol-i2c.h>
 #endif
 
+struct reader_cb_data {
+    const void *data;
+};
+
 struct sol_iio_device {
     double *mount_matrix;
     char *trigger_name;
     void (*reader_cb)(void *data, struct sol_iio_device *device);
-    const void *reader_cb_data;
+    struct sol_ptr_vector reader_cb_data_list;
```

Comment:
You also need to change `reader_cb` to a `reader_cb_list`. This list would be composed of a 
```
struct {
   void (*reader_cb)(void *data, struct sol_iio_device *device);
   const void *data;
}
```
So different callbacks (with their respective user data) can be defined. The way you've done work properly because `iio/nodes.c` has a single callback function, but on `sol-iio.c` we need to cover a more generic usage.

Minor: check if a `sol_vector` isn't enough, as you won't (probably) share list items elsewhere.